src/Tasks/Microsoft.NET.Build.Tasks/EmbedAppNameInHost.cs

```diff
@@ -0,0 +1,169 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+using System;
+using System.IO;
+using System.Text;
+using System.Linq;
+using System.Globalization;
+using System.Collections.Generic;
+using System.Security.Cryptography;
+
+namespace Microsoft.NET.Build.Tasks
+{
+    /// <summary>
+    /// Embeds the App Name into the AppHost.exe  
+    /// </summary>
+    public class EmbedAppNameInHost : TaskBase
+    {
+        private string _modifiedAppHostPath;
+
+        [Required]
+        public string AppHostSourcePath { get; set; }
+
+        [Required]
+        public string AppHostDestinationDirectoryPath { get; set; }
+
+        [Required]
+        public string AppBinaryName { get; set; }
+
+        [Output]
+        public string ModifiedAppHostPath
+        {
+            get { return _modifiedAppHostPath; }
+        }
+
+        protected override void ExecuteCore()
+        {
+            
+            var hostExtension = Path.GetExtension(AppHostSourcePath);
+           
+            var appbaseName = Path.GetFileNameWithoutExtension(AppBinaryName);
+            var bytesToWrite = Encoding.UTF8.GetBytes(AppBinaryName);
+            var destinationDirectory = Path.GetFullPath(AppHostDestinationDirectoryPath);
+            _modifiedAppHostPath = Path.Combine(destinationDirectory, $"{appbaseName}{hostExtension}");
+
+            if ( File.Exists(_modifiedAppHostPath))
+            {
+                //We have already done the required modification to apphost.exe
+                return;
+            }
+
+            if (bytesToWrite.Length > 1024)
+            {
+                throw new BuildErrorException(Strings.FileNameIsTooLong, AppBinaryName);
+            }
+
+            var array = File.ReadAllBytes(AppHostSourcePath);
+            var placeHolder = "c3ab8ff13720e8ad9047dd39466b3c8974e592c2fa383d4a3960714caef0c4f2"; //hash value embedded in apphost.exe
+            var bytesToSearch = Encoding.UTF8.GetBytes(placeHolder);
+
+            SearchAndReplace(array, bytesToSearch, bytesToWrite);
+
+
+            if ( !Directory.Exists(destinationDirectory))
+            {
+                Directory.CreateDirectory(destinationDirectory);
+            }
+            
+            File.WriteAllBytes(_modifiedAppHostPath, array);
+
+        }
+
+        // See: https://en.wikipedia.org/wiki/Knuth%E2%80%93Morris%E2%80%93Pratt_algorithm
+        static int[] ComputeKMP_FailureFunction(byte[] pattern)
```

Comment:
(We should only do this if we are super worried about perf)

Since we are always searching for the same hard-coded string, it would be more performant to also hard-code the "FailureFunction" array as well.  That way we wouldn't need to compute the same thing every time.